File: src/services/vendor-attachment.service.ts
import { VendorDocumentRepository } from "@/repositories/vendor-document.repository";
import { VendorRepository } from "@/repositories/vendor.repository";
import { ALLOWED_DOCUMENT_MIME_TYPES, type VendorAttachment, type VendorAttachmentMetadata } from "@/types/vendor/vendor_attachment";
import { type VendorDocument } from "@/types/vendor/vendor_document";
import { AppError } from "@/middleware/error-handler";
import { logger } from "@/utils/logger";
import { Timestamp } from "firebase-admin/firestore";
import { Binary, ObjectId } from "mongodb";

export interface UploadAttachmentInput {
  title: string;
  description?: string;
  filename: string;
  mimeType: string;
  size: number;
  data: Buffer;
}

const ALLOWED_EXTENSIONS = [
  ".pdf",
  ".xlsx",
  ".xls",
  ".doc",
  ".docx",
  ".jpg",
  ".jpeg",
  ".png",
  ".gif",
  ".webp",
];

export class VendorAttachmentService {
  private repository: VendorDocumentRepository;
  private vendorRepository: VendorRepository;

  constructor() {
    this.repository = new VendorDocumentRepository();
    this.vendorRepository = new VendorRepository();
  }

  /**
   * Validate that the file is an allowed type (PDF, Excel, Word, Image)
   */
  private validateFile(mimeType: string, filename: string): void {
    const isMimeAllowed = ALLOWED_DOCUMENT_MIME_TYPES.includes(mimeType as any);
    const lowerFilename = filename.toLowerCase();
    const hasAllowedExtension = ALLOWED_EXTENSIONS.some((ext) => lowerFilename.endsWith(ext));

    if (!isMimeAllowed && !hasAllowedExtension) {
      throw new AppError(
        400,
        "Only PDF, Excel (.xlsx, .xls), Word (.docx, .doc), and Image (.jpg, .png, .gif, .webp) files are allowed",
        "INVALID_FILE_TYPE"
      );
    }
  }

  /**
   * Upload a new attachment for a vendor in company database
   */
  async uploadAttachment(
    companyId: string,
    vendorId: string,
    uploaderId: string,
    uploaderName: string,
    input: UploadAttachmentInput
  ): Promise<VendorAttachmentMetadata> {
    this.validateFile(input.mimeType, input.filename);

    // Check if vendor exists in company DB
    const vendor = await this.vendorRepository.findById(vendorId, companyId);
    if (!vendor) {
      throw new AppError(404, "Vendor not found", "VENDOR_NOT_FOUND");
    }

    const document: Omit<VendorDocument, "_id"> = {
      vendorId: new ObjectId(vendorId),
      title: input.title,
      description: input.description || "",
      uploadedAt: Timestamp.now(),
      uploaderId,
      uploaderName,
      filename: input.filename,
      mimeType: input.mimeType as any,
      size: input.size,
      data: new Binary(input.data),
    };

    const created = await this.repository.create(document, vendorId, companyId);

    logger.info("Vendor attachment created", {
      documentId: created._id?.toString(),
      vendorId,
      uploaderId,
      filename: input.filename,
      companyId,
    });

    const { data, ...metadata } = created;
    return metadata as unknown as VendorAttachmentMetadata;
  }

  /**
   * Get attachment by ID (with binary data)
   */
  async getAttachment(attachmentId: string, companyId?: string): Promise<VendorAttachment> {
    const document = await this.repository.findById(attachmentId, companyId);

    if (!document) {
      throw new AppError(404, `Attachment not found: ${attachmentId}`, "ATTACHMENT_NOT_FOUND");
    }

    return document as unknown as VendorAttachment;
  }

  /**
   * Get attachment metadata by ID
   */
  async getAttachmentMetadata(attachmentId: string, companyId?: string): Promise<VendorAttachmentMetadata> {
    const document = await this.repository.findById(attachmentId, companyId);

    if (!document) {
      throw new AppError(404, "Attachment not found", "ATTACHMENT_NOT_FOUND");
    }

    const { data, ...metadata } = document;
    return metadata as unknown as VendorAttachmentMetadata;
  }

  /**
   * List all attachments for a vendor in company database
   */
  async listAttachmentsByVendor(vendorId: string, companyId?: string): Promise<VendorAttachmentMetadata[]> {
    const vendor = await this.vendorRepository.findById(vendorId, companyId);
    if (!vendor) {
      throw new AppError(404, "Vendor not found", "VENDOR_NOT_FOUND");
    }

    const documents = await this.repository.findAllByVendorId(vendorId, companyId);
    return documents.map((doc) => doc as unknown as VendorAttachmentMetadata);
  }

  /**
   * Delete an attachment
   */
  async deleteAttachment(attachmentId: string, companyId?: string): Promise<void> {
    const deleted = await this.repository.delete(attachmentId, companyId);

    if (!deleted) {
      throw new AppError(404, "Attachment/Document not found", "ATTACHMENT_NOT_FOUND");
    }

    logger.info("Vendor attachment deleted", { attachmentId, companyId });
  }

  /**
   * Delete all attachments for a vendor
   */
  async deleteAllAttachmentsForVendor(vendorId: string, companyId?: string): Promise<number> {
    return await this.repository.deleteByVendorId(vendorId, companyId);
  }
}

File: fisoft/src/main/kotlin/services/VendorAttachmentService.kt
package com.ilyasdemirkiran.services

import com.ilyasdemirkiran.middleware.AppError
import com.ilyasdemirkiran.repository.VendorAttachmentRepository
import com.ilyasdemirkiran.repository.VendorRepository
import com.ilyasdemirkiran.types.UserRole
import com.ilyasdemirkiran.types.vendors.VendorDocument
import io.ktor.http.*
import java.time.Instant

class VendorAttachmentService(
    private val repo: VendorAttachmentRepository = VendorAttachmentRepository(),
    private val vendorRepo: VendorRepository = VendorRepository()
) {
    private fun assertSudo(role: UserRole) {
        if (role != UserRole.sudo) {
            throw AppError(HttpStatusCode.Forbidden, "Only sudo users can perform this operation", "FORBIDDEN")
        }
    }

    suspend fun listAttachmentsByVendor(vendorId: String): List<VendorDocument> {
        return repo.findByVendorId(vendorId)
    }

    suspend fun getAttachmentMetadata(attachmentId: String): VendorDocument {
        return repo.findById(attachmentId)
            ?: throw AppError(HttpStatusCode.NotFound, "Attachment not found", "ATTACHMENT_NOT_FOUND")
    }

    suspend fun getAttachmentFile(attachmentId: String): Triple<ByteArray, String, String> {
        val doc = getAttachmentMetadata(attachmentId)
        val fileId = doc.fileId
            ?: throw AppError(HttpStatusCode.NotFound, "Attachment file data not found", "ATTACHMENT_NOT_FOUND")
        val (data, mimeType) = repo.downloadFromGridFS(fileId)
        val filename = doc.fileName ?: "attachment.pdf"
        return Triple(data, mimeType, filename)
    }

    suspend fun uploadAttachment(
        role: UserRole,
        vendorId: String,
        title: String,
        filename: String,
        data: ByteArray,
        mimeType: String
    ): VendorDocument {
        assertSudo(role)

        if (vendorRepo.findById(vendorId) == null) {
            throw AppError(HttpStatusCode.NotFound, "Vendor not found", "VENDOR_NOT_FOUND")
        }

        val fileId = repo.uploadToGridFS(filename, data, mimeType)

        val doc = VendorDocument(
            vendorId = vendorId,
            title = title,
            fileId = fileId,
            fileName = filename,
            fileType = mimeType,
            uploadedAt = Instant.now().toString()
        )

        return repo.create(doc)
    }

    suspend fun updateAttachment(role: UserRole, attachmentId: String, title: String?, description: String?): VendorDocument {
        assertSudo(role)
        return repo.update(attachmentId, title, description)
            ?: throw AppError(HttpStatusCode.NotFound, "Attachment not found", "ATTACHMENT_NOT_FOUND")
    }

    suspend fun deleteAttachment(role: UserRole, attachmentId: String) {
        assertSudo(role)
        if (!repo.delete(attachmentId)) {
            throw AppError(HttpStatusCode.NotFound, "Attachment not found", "ATTACHMENT_NOT_FOUND")
        }
    }
}
